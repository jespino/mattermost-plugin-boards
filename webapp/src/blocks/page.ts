// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Block, createBlock} from './block'

type PageFields = {
    contentOrder: Array<string | string[]>
    properties: Record<string, any>
    icon: string
}

type Page = Block & {
    fields: PageFields
}

function createPage(block?: Block): Page {
    const contentOrder: Array<string|string[]> = []
    const contentIds = block?.fields?.contentOrder?.filter((id: any) => id !== null)

    if (contentIds?.length > 0) {
        for (const contentId of contentIds) {
            if (typeof contentId === 'string') {
                contentOrder.push(contentId)
            } else {
                contentOrder.push(contentId.slice())
            }
        }
    }

    return {
        ...createBlock(block),
        type: 'page',
        fields: {
            contentOrder,
            icon: block?.fields?.icon,
            properties: {...block?.fields?.properties},
        },
    }
}

function sortPagesAlphabetically(pages: Page[]): Page[] {
    // Strip leading emoji to prevent unintuitive results
    return pages.map((v) => {
        return {page: v, title: v.title.replace(/^\p{Emoji}*\s*/u, '')}
    }).sort((v1, v2) => v1.title.localeCompare(v2.title)).map((v) => v.page)
}

export {Page, sortPagesAlphabetically, createPage}
