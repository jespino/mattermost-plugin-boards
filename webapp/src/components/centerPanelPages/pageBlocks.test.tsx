// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react'
import {mocked} from 'jest-mock'
import {Provider as ReduxProvider} from 'react-redux'
import {render, screen} from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import {mockDOM, mockStateStore, wrapDNDIntl} from '../../testUtils'
import {TestBlockFactory} from '../../test/testBlockFactory'
import {createTextBlock} from '../../blocks/textBlock'
import {IUser} from '../../user'
import octoClient from '../../octoClient'
import {Constants} from '../../constants'

import PageBlocks from './pageBlocks'
Object.defineProperty(Constants, 'versionString', {value: '1.0.0'})
jest.mock('react-router-dom', () => {
    const originalModule = jest.requireActual('react-router-dom')

    return {
        ...originalModule,
        useRouteMatch: jest.fn(() => {
            return {url: '/board/view'}
        }),
    }
})
jest.mock('../../octoClient')
jest.mock('../../mutator')
jest.mock('../../telemetry/telemetryClient')
jest.mock('draft-js/lib/generateRandomKey', () => () => '123')
const mockedOctoClient = mocked(octoClient, true)
describe('components/pageBlocks', () => {
    const board = TestBlockFactory.createBoard()
    board.id = '1'
    board.teamId = 'team-id'
    const activePage = TestBlockFactory.createPage(board)
    activePage.id = '1'

    const content1 = createTextBlock()
    content1.title = 'text content 1'
    content1.boardId = board.id
    content1.parentId = activePage.id
    const content2 = createTextBlock()
    content2.title = 'text content 2'
    content2.boardId = board.id
    content2.parentId = activePage.id
    const content3 = createTextBlock()
    content3.title = 'text content 3'
    content3.boardId = board.id
    content3.parentId = activePage.id

    activePage.fields.contentOrder = [content1.id, content2.id, content3.id]

    const state = {
        clientConfig: {
            value: {
                featureFlags: {
                    subscriptions: true,
                },
            },
        },
        searchText: '',
        users: {
            me: {
                id: 'user_id_1',
            },
            myConfig: {
                onboardingTourStarted: {value: false},
            },
            boardUsers: {
                'user-id-1': {username: 'username_1'},
            },
            blockSubscriptions: [],
        },
        teams: {
            current: {id: 'team-id'},
        },
        boards: {
            current: board.id,
            boards: {
                [board.id]: board,
            },
            templates: [],
            myBoardMemberships: {
                [board.id]: {userId: 'user_id_1', schemeAdmin: true},
            },
        },
        limits: {
            limits: {
                cards: 0,
                used_cards: 0,
                card_limit_timestamp: 0,
                views: 0,
            },
        },
        cards: {
            cards: [],
        },
        views: {
            views: {},
        },
        pages: {
            pages: {[activePage.id]: activePage},
            current: activePage.id,
        },
        contents: {
            contents: [content1, content2, content3],
            contentsByCard: {},
            contentsByPage: {[activePage.id]: [content1, content2, content3]},
        },
        comments: {
            comments: [],
            commentsByCard: {},
        },
        imits: {
            limits: {
                views: 0,
            },
        },
    }
    mockedOctoClient.searchTeamUsers.mockResolvedValue(Object.values(state.users.boardUsers) as IUser[])
    const store = mockStateStore([], state)
    beforeAll(() => {
        mockDOM()
        console.error = jest.fn()
    })
    beforeEach(() => {
        jest.clearAllMocks()
    })
    test('should match snapshot', () => {
        const {container} = render(wrapDNDIntl(
            <ReduxProvider store={store}>
                <PageBlocks
                    board={board}
                    activePage={activePage}
                    readonly={false}
                    canEditBoardCards={true}
                />
            </ReduxProvider>,
        ))
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot after click one content block', () => {
        const {container} = render(wrapDNDIntl(
            <ReduxProvider store={store}>
                <PageBlocks
                    board={board}
                    activePage={activePage}
                    readonly={false}
                    canEditBoardCards={true}
                />
            </ReduxProvider>,
        ))
        const boardLink = screen.getAllByTestId('block-content')[0]
        userEvent.click(boardLink)
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot when readonly', () => {
        const {container} = render(wrapDNDIntl(
            <ReduxProvider store={store}>
                <PageBlocks
                    board={board}
                    activePage={activePage}
                    readonly={true}
                    canEditBoardCards={true}
                />
            </ReduxProvider>,
        ))
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot without edit board cards permission', () => {
        const {container} = render(wrapDNDIntl(
            <ReduxProvider store={store}>
                <PageBlocks
                    board={board}
                    activePage={activePage}
                    readonly={false}
                    canEditBoardCards={false}
                />
            </ReduxProvider>,
        ))
        expect(container).toMatchSnapshot()
    })
})
