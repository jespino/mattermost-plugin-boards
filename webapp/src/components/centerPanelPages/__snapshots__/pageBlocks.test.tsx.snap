// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pageBlocks should match snapshot 1`] = `
<div>
  <div
    class="PageBlocks"
  >
    <div
      class="BlocksEditor"
    >
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="block-actions"
            data-testid="block-actions"
          >
            <button
              class="IconButton size--small"
              type="button"
            >
              <i
                class="CompassIcon icon-plus"
              />
            </button>
            <div
              draggable="true"
            >
              <button
                class="IconButton size--small"
                type="button"
              >
                <i
                  class="CompassIcon icon-drag-vertical"
                />
              </button>
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 1
              </p>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="block-actions"
            data-testid="block-actions"
          >
            <button
              class="IconButton size--small"
              type="button"
            >
              <i
                class="CompassIcon icon-plus"
              />
            </button>
            <div
              draggable="true"
            >
              <button
                class="IconButton size--small"
                type="button"
              >
                <i
                  class="CompassIcon icon-drag-vertical"
                />
              </button>
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 2
              </p>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="block-actions"
            data-testid="block-actions"
          >
            <button
              class="IconButton size--small"
              type="button"
            >
              <i
                class="CompassIcon icon-plus"
              />
            </button>
            <div
              draggable="true"
            >
              <button
                class="IconButton size--small"
                type="button"
              >
                <i
                  class="CompassIcon icon-drag-vertical"
                />
              </button>
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 3
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="Editor"
      >
        <div
          class="RootInput css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class="RootInput__control css-9fsvdt-control"
          >
            <div
              class="RootInput__value-container css-30zlo3-ValueContainer"
            >
              <div
                class="RootInput__placeholder css-vinljo-placeholder"
                id="react-select-2-placeholder"
              >
                Add text or type "/" for commands
              </div>
              <div
                class="RootInput__input-container css-g5309v-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-2-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class="RootInput__input"
                  id="react-select-2-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="RootInput__indicators css-1hb7zxy-IndicatorsContainer"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/pageBlocks should match snapshot after click one content block 1`] = `
<div>
  <div
    class="PageBlocks"
  >
    <div
      class="BlocksEditor"
    >
      <div>
        <div
          class="Editor"
        >
          <div
            class="TextContent"
            data-testid="text"
          >
            <div
              class="MarkdownEditor octo-editor  active"
            >
              <div
                class="MarkdownEditorInput"
              >
                <div
                  class="DraftEditor-root"
                >
                  <div
                    class="DraftEditor-editorContainer"
                  >
                    <div
                      aria-autocomplete="list"
                      aria-expanded="false"
                      class="notranslate public-DraftEditor-content"
                      contenteditable="true"
                      role="combobox"
                      spellcheck="false"
                      style="outline: none; user-select: text; white-space: pre-wrap; word-wrap: break-word;"
                    >
                      <div
                        data-contents="true"
                      >
                        <div
                          class=""
                          data-block="true"
                          data-editor="123"
                          data-offset-key="123-0-0"
                        >
                          <div
                            class="public-DraftStyleDefault-block public-DraftStyleDefault-ltr"
                            data-offset-key="123-0-0"
                          >
                            <span
                              data-offset-key="123-0-0"
                            >
                              <span
                                data-text="true"
                              >
                                text content 1
                              </span>
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="block-actions"
            data-testid="block-actions"
          >
            <button
              class="IconButton size--small"
              type="button"
            >
              <i
                class="CompassIcon icon-plus"
              />
            </button>
            <div
              draggable="true"
            >
              <button
                class="IconButton size--small"
                type="button"
              >
                <i
                  class="CompassIcon icon-drag-vertical"
                />
              </button>
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 2
              </p>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="block-actions"
            data-testid="block-actions"
          >
            <button
              class="IconButton size--small"
              type="button"
            >
              <i
                class="CompassIcon icon-plus"
              />
            </button>
            <div
              draggable="true"
            >
              <button
                class="IconButton size--small"
                type="button"
              >
                <i
                  class="CompassIcon icon-drag-vertical"
                />
              </button>
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 3
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/pageBlocks should match snapshot when readonly 1`] = `
<div>
  <div
    class="PageBlocks"
  >
    <div
      class="BlocksEditor"
    >
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 1
              </p>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 2
              </p>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 3
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/pageBlocks should match snapshot without edit board cards permission 1`] = `
<div>
  <div
    class="PageBlocks"
  >
    <div
      class="BlocksEditor"
    >
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 1
              </p>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 2
              </p>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="BlockContent   "
          data-testid="block-content"
          style="opacity: 1;"
        >
          <div
            class="content"
          >
            <div
              class="octo-editor-preview"
            >
              <p>
                text content 3
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
